File: ios/Classes/Video/VideoAd.swift
//
//  VideoAd.swift
//  nend_plugin
//
//

import UIKit
import Flutter
import NendAd

class VideoAd: AdBridger {
    
    var video: NADVideo!
    
    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch targetMethod(call.method) {
            case .isReady: return result(video.isReady)
            case .loadAd: loadAd()
            case .showAd: showAd()
            case .releaseAd: video.releaseAd()
            case .mediationName:
                video.mediationName = videoCodable(from: call.arguments)?.mediationName ?? ""
            default:
                return
        }
        result(true)
    }
    
    func loadAd() {
        if video != nil {
            video.loadAd()
        }
    }
    
    private func showAd() {
        guard let rootViewController = self.rootViewController else { return }
        video.showAd(from: rootViewController)
    }
    
    func adUnit(from argument: Any?) -> AdUnit {
        let jsonData = try! JSONSerialization.data(withJSONObject: argument!, options: [])
        return try! JSONDecoder().decode(AdUnit.self, from: jsonData)
    }
    
    func videoCodable(from argument: Any?) -> VideoCodable? {
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: argument!, options: [])
            return try JSONDecoder().decode(VideoCodable.self, from: jsonData)
        } catch let error {
            print(error)
        }
        return nil
    }
}

File: android/src/main/kotlin/net/nend/nendplugin/VideoAd.kt
package net.nend.nendplugin

import android.app.Activity
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result
import net.nend.android.NendAdVideo
import org.json.JSONObject

abstract class VideoAd<Ad : NendAdVideo>(private val activity: Activity?) : AdBridger() {

    var video: Ad? = null

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (targetMethod(call.method)) {
            MethodName.IsReady -> {
                return result.success(video?.isLoaded)
            }
            MethodName.LoadAd -> video?.loadAd()
            MethodName.ShowAd -> video?.showAd(activity)
            MethodName.ReleaseAd -> video?.releaseAd()
            MethodName.MediationName -> video?.setMediationName(call.argument("mediationName"))
            else -> return
        }
        result.success(true)
    }

    abstract fun initAd(adUnit: AdUnit?)

}
